usr/lib/common/pin_get.c

```diff
@@ -0,0 +1,114 @@
+/*
+ * COPYRIGHT (c) International Business Machines Corp. 2022
+ *
+ * This program is provided under the terms of the Common Public License,
+ * version 1.0 (CPL-1.0). Any use, reproduction or distribution for this
+ * software constitutes recipient's acceptance of CPL-1.0 terms which can be
+ * found in the file LICENSE file or at
+ * https://opensource.org/licenses/cpl1.0.php
+ */
+
+#define _POSIX_C_SOURCE 200809L
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <termios.h>
+#include <unistd.h>
+#include <sys/mman.h>
+#include <openssl/crypto.h>
+
+#include <pin_get.h>
+
+static void echo(bool on)
+{
+    struct termios term;
+
+    tcgetattr(fileno(stdin), &term);
+
+    if (on)
+        term.c_lflag |= ECHO;
+    else
+        term.c_lflag &= ~ECHO;
+
+    tcsetattr(fileno(stdin), TCSAFLUSH, &term);
+}
+
+void pin_free(char **buf)
+{
+    if (!buf)
+        return;
+
+    if (*buf)
+        OPENSSL_cleanse(*buf, strlen(*buf));
+
+    free(*buf);
+    *buf = NULL;
+}
+
+const char *pin_prompt(char **buf, const char *msg)
+{
+    ssize_t n;
+    size_t s;
+
+    if (!buf)
+        return NULL;
+
+    printf("%s", msg);
+    fflush(stdout);
+    echo(false);
+
+    n = getline(buf, &s, stdin);
```

Comment:
I guess you need to take care about the situation that *buf is non-NULL on entry of the pin_prompt funmction.
getline accepts a non-null lineptr argument (aka buff), but then expects *n (aka s) to comtain the size if the buffer in *lineptr.

So either make sure that *buf can not be non-NULL when pin_prompt is called, or setup s like `s = strlen(*buf) + 1` when *buf is non NULL. Alternatively, just free *buf if it is not NULL, and let getline allocate a new one.

Reply:
Since we're no longer using a struct, *buf != NULL is a coding error. So I would prefer to return NULL in these cases.